### ISSUE TITLE
[BUG] deepspeed.utils.safe_get_full_grad get all nan value

### ISSUE THREAD

**Issue by gin-xi** [microsoft/DeepSpeed#5883]
**Describe the bug**
hello, i try to train model with deepspeed ZERO, this is my script
```
import os
import torch
import deepspeed
from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig
from torch.utils.data import DataLoader, TensorDataset

# Constants
MODEL_NAME = "facebook/opt-125m"
MAX_LENGTH = 512
PER_DEVICE_TRAIN_BATCH_SIZE = 2
GRADIENT_ACCUMULATION_STEPS = 1
NUM_EPOCHS = 1
LEARNING_RATE = 5e-5
WEIGHT_DECAY = 0.01
LOGGING_STEPS = 10
MAX_STEPS = 100

def main():
    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
    tokenizer.pad_token = tokenizer.eos_token

    # Creating a dummy dataset
    dummy_data = torch.randint(0, tokenizer.vocab_size, (1000, MAX_LENGTH))
    dummy_labels = dummy_data.clone()
    train_dataset = TensorDataset(dummy_data, dummy_labels)
    train_loader = DataLoader(train_dataset, batch_size=PER_DEVICE_TRAIN_BATCH_SIZE, shuffle=True)

    config = AutoConfig.from_pretrained(MODEL_NAME)
    model = AutoModelForCausalLM.from_config(config)
    model.gradient_checkpointing_enable()

    optimizer = torch.optim.AdamW(model.parameters(), lr=LEARNING_RATE, weight_decay=WEIGHT_DECAY)

    model_engine, optimizer, _, _ = deepspeed.initialize(
        model=model,
        optimizer=optimizer,
        config_params={
            "train_batch_size": 4,
            "train_micro_batch_size_per_gpu": PER_DEVICE_TRAIN_BATCH_SIZE,
            "gradient_accumulation_steps": GRADIENT_ACCUMULATION_STEPS,
            "fp16": {
                "enabled": True
            },
            "zero_optimization": {
                "stage": 1
            }
        }
    )

    for epoch in range(NUM_EPOCHS):
        model_engine.train()
        for step, batch in enumerate(train_loader):
            inputs = batch[0].to(model_engine.local_rank)
            labels = batch[1].to(model_engine.local_rank)
            outputs = model_engine(input_ids=inputs, labels=labels)
            loss = outputs.loss
            model_engine.backward(loss)
            model_engine.step()

            if step % LOGGING_STEPS == 0:
                print(f"Step: {step}, Loss: {loss.item()}")
                for name, param in model.named_parameters():
                    grad = deepspeed.utils.safe_get_full_grad(param)
                    print(f"{name} grad: {grad}")
                    print(f"grad sum: {grad.nan_to_num(0).sum()}")
            if step >= MAX_STEPS:
                break

if __name__ == "__main__":
    torch.manual_seed(0)
    main()
```

i try to get the grad value for analysis, but get almost all nan via **deepspeed.utils.safe_get_full_grad** like the following image.
<img width="879" alt="image" src="https://github.com/user-attachments/assets/f69fe2af-515f-47e0-b01d-994cbe91f953">

**To Reproduce**
1. run by `torchrun --nproc_per_node=2 train.py` or `deepspeed --num_gpus=2 train.py`

**Expected behavior**
expect some normal value for gradients.
**ds_report output**
<img width="776" alt="image" src="https://github.com/user-attachments/assets/9b40c72d-9c2a-4546-8168-8bd4892799de">
**Screenshots**
<img width="908" alt="image" src="https://github.com/user-attachments/assets/90b49ec8-d47e-4232-92ed-7d858f7265ce">
**System info (please complete the following information):**
 - nvcr.io/nvidia/cuda  12.4.1-cudnn-devel-ubuntu22.04
 - single machine with x8 H100s

**Launcher context**
`deepspeed` launcher or `torchrun` both have same error

**Comment by jomayeri** [comment]
It appears to be an instability in your fp16 training that is causing the `nan`s. When I switch to bf16 the issue went away. Also please change your code so that the print is after `backward` but before `step` like this 

```Python
    for epoch in range(NUM_EPOCHS):
        model_engine.train()
        for step, batch in enumerate(train_loader):
            inputs = batch[0].to(model_engine.local_rank)
            labels = batch[1].to(model_engine.local_rank)
            outputs = model_engine(input_ids=inputs, labels=labels)
            loss = outputs.loss
            model_engine.backward(loss)
            if step % LOGGING_STEPS == 0:
                 print(f"Step: {step}, Loss: {loss.item()}")
                 for name, param in model.named_parameters():
                     grad = deepspeed.utils.safe_get_full_grad(param)
                     print(f"{name} grad: {grad}")
                     print(f"grad sum: {grad.nan_to_num(0).sum()}")
            model_engine.step()
            if step >= MAX_STEPS:
                break
```

**Comment by jing-4369** [comment]
> It appears to be an instability in your fp16 training that is causing the `nan`s. When I switch to bf16 the issue went away. Also please change your code so that the print is after `backward` but before `step` like this
> 
> ```python
>     for epoch in range(NUM_EPOCHS):
>         model_engine.train()
>         for step, batch in enumerate(train_loader):
>             inputs = batch[0].to(model_engine.local_rank)
>             labels = batch[1].to(model_engine.local_rank)
>             outputs = model_engine(input_ids=inputs, labels=labels)
>             loss = outputs.loss
>             model_engine.backward(loss)
>             if step % LOGGING_STEPS == 0:
>                  print(f"Step: {step}, Loss: {loss.item()}")
>                  for name, param in model.named_parameters():
>                      grad = deepspeed.utils.safe_get_full_grad(param)
>                      print(f"{name} grad: {grad}")
>                      print(f"grad sum: {grad.nan_to_num(0).sum()}")
>             model_engine.step()
>             if step >= MAX_STEPS:
>                 break
> ```

thank you! That definitely fixed the problem
